### ISSUE TITLE
Can't get Scala to build

### ISSUE THREAD

**Issue by soapergem** [thought-machine/pleasings#25]
TL;DR: **How do I get Scala code buildling with Please?**

Today I tried making use of the custom Scala rules from this repository using Please 13.4.6 but have been running into a lot of trouble just getting a simple Jar to build. To give you an idea of what I'm working with, here's the package structure (roughly):

```
.
├── build_defs
│   └── scala.build_defs
├── src
│   └── main
│       ├── resources
│       │   └── Log4j.properties
│       └── scala
│           └── com
│               └── mypackage
│                   └── spark
│                       ├── BUILD
│                       ├── MyHelperClass.scala
│                       └── MyMainClass.scala
├── third_party
│   └── scala
│       └── BUILD
├── .plzconfig
└── pom.xml
```

I copied the `build_defs/scala.build_defs` file verbatim from [scala/scala.build_defs](https://github.com/thought-machine/pleasings/blob/master/scala/scala.build_defs). At first I tried to run this command:

```
please build -o parse.preloadbuilddefs:build_defs/scala.build_defs
```

But that died almost immediately with this error:

```
CRITICAL: Error loading builtin rules: build_defs/scala.build_defs:184:1: error: Config object is not assignable in this scope

CONFIG.setdefault('SCALAC_TOOL', 'scalac')
```

So, next I tried removing those three lines at the bottom of the build_defs file, and updated my `.plzconfig` file to look like this:

```
[scala]
scalacTool = scalac
scalaLibrary = //third_party/scala:scala_library
scalaVersion = 2.11
```

Then when I ran the same build command, I got this somewhat longer error:

```
WARNING: Error in config file: warnings:
can't store data at section "scala"
can't store data at section "scala"
can't store data at section "scala"
can't store data at section "scala"
can't store data at section "scala"
can't store data at section "scala"
can't store data at section "scala"
can't store data at section "scala"

Build stopped after 50ms. 1 target failed:
    //src/main/scala/com/mypackage/spark:all
build_defs/scala.build_defs:54:21: error: Config has no such property SCALA_LIBRARY

        deps.append(CONFIG.SCALA_LIBRARY)
                    ^


Traceback:
                                                build_defs/scala.build_defs:54:21:           deps.append(CONFIG.SCALA_LIBRARY)
                                                                           : 0: 0:   <source unavailable>
                                                build_defs/scala.build_defs:54: 9:           deps.append(CONFIG.SCALA_LIBRARY)
                                                build_defs/scala.build_defs:29: 5:       if srcs or src_dir:
/home/username/myproject/src/main/scala/com/mypackage/spark/BUILD: 3: 1:   scala_library(
```

The next thing I tried was updating my `.plzconfig` file to this:

```
[scala]
scalacTool = scalac
scalaLibrary = //third_party/scala:scala_library
scalaVersion = 2.11

[build]
passenv = SCALAC_TOOL
passenv = SCALA_LIBRARY
passenv = SCALA_VERSION
```

Then I tried setting each of those values as environment variables first, and then running the same build command. However even that gave me this error message:

```
Build stopped after 50ms. 2 targets failed:
    //src/main/scala/com/mypackage/spark:all
build_defs/scala.build_defs:54:21: error: Config has no such property SCALA_LIBRARY

        deps.append(CONFIG.SCALA_LIBRARY)
                    ^


Traceback:
                                                build_defs/scala.build_defs:54:21:           deps.append(CONFIG.SCALA_LIBRARY)
                                                                           : 0: 0:   <source unavailable>
                                                build_defs/scala.build_defs:54: 9:           deps.append(CONFIG.SCALA_LIBRARY)
                                                build_defs/scala.build_defs:29: 5:       if srcs or src_dir:
/home/username/myproject/src/main/scala/com/mypackage/spark/BUILD: 3: 1:   scala_library(
    //third_party/scala:all
build_defs/scala.build_defs:174:14: error: Config has no such property SCALA_VERSION
        name = name,
        id = f'{org}:{artifact}_%s:{version}' % CONFIG.SCALA_VERSION,
             ^
        hash = hash,

Traceback:
                                build_defs/scala.build_defs:174:14:           id = f'{org}:{artifact}_%s:{version}' % CONFIG.SCALA_VERSION,
                                build_defs/scala.build_defs:172:12:       return maven_jar(
/home/username/myproject/third_party/scala/BUILD: 17: 1:   scala_jar(
```

I have no idea at this point how to get this working. Any advice?

**Comment by soapergem** [comment]
@peterebden sorry to bug you directly but do you know if the scala rules should work as-is? Any clues to what I'm doing wrong re: the config vars?

**Comment by peterebden** [comment]
Okay, so the "Config object is not assignable" error looks to be related to trying to preload it. I'll have a look into that; not immediately sure offhand why that happens, but it is solvable by using a `subinclude` rather than the preload mechanism.

On the config, there isn't a separate section for `scala`, the config library we use doesn't really support that. Instead you can configure it like so:
```
[buildconfig]
scalac-tool = scalac
scala-library = //third_party/scala:scala-library
```

Hope that helps!

**Comment by soapergem** [comment]
This was very helpful, thank you! However now I've run into an entirely separate issue that still fits the title of the issue (i.e. "Can't get Scala to build"). More accurately, I _can_ get things building now; however it's not building enough. I confess I don't really know much about what the scalac compiler does so I'm about to make some big assumptions...

I have a main class that makes use of the [Scopt library](https://github.com/scopt/scopt) for parsing command line arguments. The syntax that Scopt uses makes heavy use of anonymous functions, and a syntax whose name I don't even know the proper name. Here's my main class:

```
package com.mypackage.spark

case class Config(input: String = null)

object MyMainClass {

  val parser = new scopt.OptionParser[Config]("testapp") {
    arg[String]("input") action { (x, c) => c.copy(input = x) } text("input value")
  }

  def main(args: Array[String]) {
    parser.parse(args, Config()) map { config =>
      val input = config.input
      println("input value is " + input)
    } getOrElse {
      System.exit(1)
    }
  }
}
```

When I build the code using Maven, under com.mypackage.spark, I can see that it generates the following eight class files (inside the JAR file):

* Config.class
* Config$.class
* Main$$anon$1$$anonfun$1.class
* Main$$anon$1.class
* Main$$anonfun$main$1.class
* Main$$anonfun$main$2.class
* Main.class
* Main$.class

However the JAR file that Please builds only includes the following five of those classes:

* Config.class
* Config$.class
* Main$$anon$1.class
* Main.class
* Main$.class

In other words, it does not generate the three anonfun classes. I'm not sure which is to blame: jarcat or scalac. Unfortunately, even with verbosity turned on, Maven isn't very forthcoming about which subprocesses it is calling under the hood, so I'm not sure if there's some flag for scalac that needs to be set. I was hoping that by posting the code snippet above it might spark some inspiration for you? It seems like I am very close to having this building, but I'm getting stumped as to why Please seemingly isn't building those last three class files.

**Comment by soapergem** [comment]
In fact, I've just made [a sample repo](https://github.com/soapergem/scala-please-test) showcasing the problem I'm running into, in case you wanted to see exactly what I'm doing.

**Comment by soapergem** [comment]
Nevermind, I think it I figured out the problem! I was trying to build Scala 2.11 code but the version of `scalac` that I had Please pointed at was 2.12. Thanks again for your help earlier.
